### ISSUE TITLE
Unexpected pyright errors (only in CLI, not in pylance)

### ISSUE THREAD

**Issue by karolzlot** [microsoft/pyright#4784]
**Describe the bug**
When using pylance I don't see any error.
When using pyright CLI I'm getting errors.

**To Reproduce**
Use code from below.
(Let me know if more information is needed, I will try to provide it)

**Expected behavior**
I think this code should have no errors in type checking. (both in pylance and in pyright CLI)

**Screenshots or Code**

![image](https://user-images.githubusercontent.com/11590960/225618821-e25c0119-61d8-46a2-9f85-e93b9a9fbaa5.png)

```python
from pandas import DataFrame
from typing import Literal, Optional, Union

def worker(
    batchsize: int,
):
   df : Optional[Union[DataFrame, Literal[-1]]] = DataFrame()

   while True:
        assert isinstance(df, DataFrame)
        if len(df.index) >= batchsize:

            df = df[batchsize:]
        else:
            break
```

`pip freeze`:
```
nodeenv==1.7.0
numpy==1.24.2
pandas==1.5.3
pyright==1.1.299
python-dateutil==2.8.2
pytz==2022.7.1
six==1.16.0
```

`requirements.txt`:
```
pyright
pandas
```

**VS Code extension or command-line**
pylance v2023.3.20

pyright CLI -> I tried two versions:

1.1.298:
```
No configuration file found.
No pyproject.toml file found.
Assuming Python platform Linux
Searching for source files
Found 1 source file
pyright 1.1.298
/home/karol/Desktop/projekty/pyright_issue1/example.py
  /home/karol/Desktop/projekty/pyright_issue1/example.py:13:18 - error: "__getitem__" method not defined on type "Literal[-1]" (reportGeneralTypeIssues)
  /home/karol/Desktop/projekty/pyright_issue1/example.py:11:19 - error: "index" is not a known member of "None" (reportOptionalMemberAccess)
  /home/karol/Desktop/projekty/pyright_issue1/example.py:11:19 - error: Cannot access member "index" for type "Literal[-1]"
    Member "index" is unknown (reportGeneralTypeIssues)
  /home/karol/Desktop/projekty/pyright_issue1/example.py:13:18 - error: Object of type "None" is not subscriptable (reportOptionalSubscript)
4 errors, 0 warnings, 0 informations 
Completed in 0.618sec
WARNING: there is a new pyright version available (v1.1.298 -> v1.1.299).
Please install the new version or set PYRIGHT_PYTHON_FORCE_VERSION to `latest`
```

1.1.299:
```
No configuration file found.
No pyproject.toml file found.
Assuming Python platform Linux
Searching for source files
Found 1 source file
pyright 1.1.299
/home/karol/Desktop/projekty/pyright_issue1/example.py
  /home/karol/Desktop/projekty/pyright_issue1/example.py:13:18 - error: "__getitem__" method not defined on type "Literal[-1]" (reportGeneralTypeIssues)
  /home/karol/Desktop/projekty/pyright_issue1/example.py:11:19 - error: "index" is not a known member of "None" (reportOptionalMemberAccess)
  /home/karol/Desktop/projekty/pyright_issue1/example.py:11:19 - error: Cannot access member "index" for type "Literal[-1]"
    Member "index" is unknown (reportGeneralTypeIssues)
  /home/karol/Desktop/projekty/pyright_issue1/example.py:13:18 - error: Object of type "None" is not subscriptable (reportOptionalSubscript)
4 errors, 0 warnings, 0 informations 
Completed in 0.589sec
```

**Additional context**
n/a

**Comment by erictraut** [comment]
The differences are likely due to one of the following:
1. Pylance bundles some additional type stubs that pyright does not. These type stubs can be found [here](https://github.com/microsoft/python-type-stubs) if you want to install them in your local "typings" folder within your project. That way, they will be available to both pylance and pyright.
2. Pylance defaults `useLibraryCodeForTypes` to true whereas pyright defaults it to false. This is because pylance users tend to be more focused on completion suggestions whereas pyright is focused on type checking.
3. There's a small possibility that the differences are caused by different versions of pyright. Pylance is built on top of pyright, but it is one week behind in terms of bug fixes, etc. The "insiders build" of pylance uses the latest version of pyright.

**Comment by karolzlot** [comment]
Thank you, it was enough to add `useLibraryCodeForTypes = true` to `pyproject.toml`.
